File: apps/android/app/src/main/java/com/kungsbackacarcommunity/app/chatchannels/FirebaseConvoyChatRepository.kt
package com.kungsbackacarcommunity.app.chatchannels

import android.content.Context
import com.google.firebase.FirebaseApp
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreException
import com.google.firebase.firestore.Query
import com.google.firebase.functions.FirebaseFunctions
import com.kungsbackacarcommunity.app.blocking.BlockVisibility
import com.kungsbackacarcommunity.app.blocking.BlockVisibilityRepository
import com.kungsbackacarcommunity.app.blocking.FirebaseBlockVisibilityRepository
import com.kungsbackacarcommunity.app.profile.FirebaseLiveProfileRepository
import com.kungsbackacarcommunity.app.profile.LiveProfileRepository
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.combine

/**
 * [ConvoyChatRepository] backed by the `convoy-list` callable (for the caller's
 * chat-eligible convoys), the accepted-member-readable
 * `convoyChats/{convoyId}/messages` listener, and the `convoyChat-*` callables
 * (europe-west1). Guarded ([createIfAvailable]) so a config-less build gets a
 * null repository and the tab renders a placeholder.
 *
 * BLOCKING: the live window is filtered against the caller's mutual-hidden set
 * ([BlockVisibilityRepository]), so two members of the same convoy who have
 * blocked each other stop seeing each other's messages in both directions.
 * Older pages are filtered SERVER-side by `convoyChat-list`; the live window
 * cannot be, because a Firestore rule cannot filter a list query per document —
 * see [BlockVisibility].
 *
 * LIVE PROFILES: as in the community channel, the sender profile stamped on each
 * message is overlaid with the sender's current `users/{uid}` profile
 * ([LiveProfileRepository], [ChannelThread.hydrate]) on both the live window and
 * older pages.
 *
 * UNREAD: [observeUnread] combines its OWN bounded newest-message listener with
 * the owner-readable `userPrivate/{uid}.convoyChatLastReadAt.{convoyId}` marker —
 * community's design, generalized from a dot to a count. Both listeners are
 * separate from [observeMessages] on purpose: the badge is shown by the map
 * shell, where the channel itself is not open, so tying the count to the
 * channel's window would mean either syncing a chat nobody has opened or no badge
 * at all.
 */
class FirebaseConvoyChatRepository private constructor(
    private val firestore: FirebaseFirestore,
    private val functions: FirebaseFunctions,
    private val blockVisibility: BlockVisibilityRepository,
    private val liveProfiles: LiveProfileRepository,
) : ConvoyChatRepository {

    override suspend fun listConvoys(): ConvoyListState =
        functions.callChannel(CONVOY_LIST, emptyMap()).fold(
            onSuccess = { ConvoyListState.Loaded(ConvoyChatMapper.chatEligibleConvoys(it)) },
            onFailure = { ConvoyListState.Error },
        )

    override fun observeMessages(convoyId: String): Flow<ChannelMessagesState> =
        combine(observeRawMessages(convoyId), blockVisibility.observeHiddenUids()) { state, hidden ->
            // One document listener supplies the hidden set for the whole
            // session; the filter itself costs no reads.
            when (state) {
                is ChannelMessagesState.Loaded ->
                    ChannelMessagesState.Loaded(
                        BlockVisibility.filterHiddenAuthors(state.messages, hidden) { it.senderUid },
                    )
                ChannelMessagesState.Loading -> state
            }
        }
            .hydrateSenders(liveProfiles)

    private fun observeRawMessages(convoyId: String): Flow<ChannelMessagesState> = callbackFlow {
        val registration =
            firestore
                .collection(CONVOY_CHATS)
                .document(convoyId)
                .collection(MESSAGES)
                .orderBy(CREATED_AT, Query.Direction.DESCENDING)
                .limit(CHANNEL_MESSAGES_PAGE_SIZE.toLong())
                .addSnapshotListener { snapshot, error ->
                    if (error != null) {
                        if ((error as? FirebaseFirestoreException)?.code ==
                            FirebaseFirestoreException.Code.PERMISSION_DENIED
                        ) {
                            // Removed from the convoy / blocked: emit the empty/denied
                            // state, clearing to no messages even when a cached
                            // snapshot exists, so denied convoy history is never
                            // shown. Applies whether or not a snapshot came back (a
                            // first-load deny also clears to empty rather than staying
                            // in Loading). PERMISSION_DENIED is terminal — no retry
                            // undoes it.
                            trySend(ChannelMessagesState.Loaded(emptyList()))
                        } else if (snapshot != null) {
                            // Transient failure (UNAVAILABLE/timeout) WITH cached
                            // data: prefer the last-known messages; the SDK retries.
                            val cached =
                                snapshot.documents.mapNotNull { it.toChannelMessage() }.asReversed()
                            trySend(ChannelMessagesState.Loaded(cached))
                        }
                        // Transient failure with NO cached data: don't emit — an
                        // empty list would misrender offline/unavailable as "no
                        // messages". The SDK retries; the initial Loading holds
                        // until a real snapshot arrives.
                        return@addSnapshotListener
                    }
                    val messages =
                        snapshot?.documents?.mapNotNull { it.toChannelMessage() }
                            ?.asReversed().orEmpty()
                    trySend(ChannelMessagesState.Loaded(messages))
                }
        awaitClose { registration.remove() }
    }

    override fun observeUnread(convoyId: String, uid: String): Flow<Int> {
        // The window is [UNREAD_SCAN_LIMIT], and it is what makes the count both
        // cheap and honest. It is ONE more than the badge's display cap
        // (ConvoyBar.UNREAD_DISPLAY_MAX), so anything at or past the cap renders
        // as "9+" whether the true backlog is 10 or 10 000 — the saturated count
        // and the real one are indistinguishable on screen, and nothing pays for a
        // count() aggregation or a wider sync to tell them apart. The documented
        // bound is that this can never render an exact number above the cap.
        val window: Flow<List<ChannelMessage>?> = callbackFlow {
            val registration =
                firestore
                    .collection(CONVOY_CHATS)
                    .document(convoyId)
                    .collection(MESSAGES)
                    .orderBy(CREATED_AT, Query.Direction.DESCENDING)
                    .limit(UNREAD_SCAN_LIMIT)
                    .addSnapshotListener { snapshot, error ->
                        if (error != null) {
                            if ((error as? FirebaseFirestoreException)?.code ==
                                FirebaseFirestoreException.Code.PERMISSION_DENIED
                            ) {
                                // Removed from the convoy / blocked: hard-clear the
                                // badge even if a stale cached snapshot is present.
                                // Never leave a count lit for a chat the caller can
                                // no longer open.
                                trySend(null)
                                return@addSnapshotListener
                            }
                            // Other (transient) error with no cached data: keep the
                            // last-known count rather than emitting a misleading
                            // zero. With cached data, fall through and use it.
                            if (snapshot == null) return@addSnapshotListener
                        }
                        trySend(snapshot?.documents?.mapNotNull { it.toChannelMessage() }.orEmpty())
                    }
            awaitClose { registration.remove() }
        }
        // Blocked authors are filtered out of the count for the same reason the
        // channel filters them out of its live window: counting a message the
        // caller will never be shown sends them into an apparently unchanged chat.
        val visible: Flow<List<ChannelMessage>> =
            combine(window, blockVisibility.observeHiddenUids()) { messages, hidden ->
                if (messages == null) {
                    emptyList()
                } else {
                    BlockVisibility.filterHiddenAuthors(messages, hidden) { it.senderUid }
                }
            }
        val lastReadAt: Flow<Long?> = callbackFlow {
            val registration =
                firestore
                    .collection(USER_PRIVATE)
                    .document(uid)
                    .addSnapshotListener { snapshot, error ->
                        if (error != null && snapshot == null) {
                            // Transient failure with no cached marker: keep the
                            // last-known marker rather than momentarily reading it
                            // as missing, which would wrongly re-light the badge.
                            return@addSnapshotListener
                        }
                        // A map keyed by convoy id (see convoyChat.markRead): an
                        // absent entry means "never opened", which counts every
                        // message from someone else.
                        val markers = snapshot?.get(CONVOY_LAST_READ_AT) as? Map<*, *>
                        trySend((markers?.get(convoyId) as? Timestamp)?.toDate()?.time)
                    }
            awaitClose { registration.remove() }
        }
        return combine(visible, lastReadAt) { messages, marker ->
            ChannelThread.unreadCount(messages, uid, marker)
        }
    }

    override fun observeAnyUnread(uid: String): Flow<Boolean> = callbackFlow {
        // ONE listener on the caller's own private document supplies BOTH maps the
        // derivation needs — the per-convoy newest-message stamps
        // (convoyChatLatestAt, maintained by the convoyChat.post fan-out) and the
        // per-convoy last-read markers (convoyChatLastReadAt, stamped by markRead).
        // No per-convoy message listener is opened, so this is a single document
        // read regardless of how many convoys the caller is in.
        val registration =
            firestore
                .collection(USER_PRIVATE)
                .document(uid)
                .addSnapshotListener { snapshot, error ->
                    if (error != null && snapshot == null) {
                        // Transient failure with no cached doc: keep the last-known
                        // value rather than momentarily reading both maps as empty,
                        // which would wrongly clear (or fail to light) the dot.
                        return@addSnapshotListener
                    }
                    val latest = timestampMillisByConvoy(snapshot?.get(CONVOY_LATEST_AT))
                    val lastRead = timestampMillisByConvoy(snapshot?.get(CONVOY_LAST_READ_AT))
                    trySend(ChannelThread.anyConvoyUnread(latest, lastRead))
                }
        awaitClose { registration.remove() }
    }

    override suspend fun markRead(convoyId: String) {
        // Best-effort idempotent bookkeeping; a transient failure is swallowed.
        functions.callChannel(MARK_READ, mapOf("convoyId" to convoyId))
    }

    override suspend fun post(convoyId: String, text: String, clientId: String?): ChannelSendResult =
        post(convoyId, text, clientId, replyToMessageId = null)

    override suspend fun post(
        convoyId: String,
        text: String,
        clientId: String?,
        replyToMessageId: String?,
    ): ChannelSendResult =
        functions.callChannel(
            POST,
            buildMap {
                put("convoyId", convoyId)
                put("text", text.trim())
                // Only sent when present, so an omitted key is a legacy auto-id doc
                // rather than a null the strict backend schema would reject.
                if (clientId != null) put("clientId", clientId)
                // Inline reply target — snapshotted server-side (and ignored while
                // chatReplies is off).
                if (replyToMessageId != null) put("replyToMessageId", replyToMessageId)
            },
        ).fold(
            onSuccess = { ChannelResponseParser.parsePostSuccess(it) },
            onFailure = {
                ChannelSendResult.Failed(ChannelErrorMapper.mapSend(it.toChannelErrorCode()))
            },
        )

    override suspend fun loadOlder(convoyId: String, before: String): ChannelOlderResult =
        functions.callChannel(LIST, mapOf("convoyId" to convoyId, "before" to before)).fold(
            onSuccess = {
                ChannelOlderResult.Loaded(
                    ChannelResponseParser.parseMessagesPage(it).hydrateSenders(liveProfiles),
                )
            },
            onFailure = { ChannelOlderResult.Failed },
        )

    companion object {
        private const val CONVOY_CHATS = "convoyChats"
        private const val MESSAGES = "messages"
        private const val CREATED_AT = "createdAt"
        private const val USER_PRIVATE = "userPrivate"
        private const val CONVOY_LAST_READ_AT = "convoyChatLastReadAt"
        private const val CONVOY_LATEST_AT = "convoyChatLatestAt"

        /**
         * Reads a `{ [convoyId]: Timestamp }` map field into `convoyId -> epoch
         * millis`, dropping any entry whose key or value is not the expected shape
         * (a legacy/partial document can never be written by the client but may
         * still hold junk). Shared by both per-convoy maps on userPrivate.
         */
        private fun timestampMillisByConvoy(raw: Any?): Map<String, Long> {
            val map = raw as? Map<*, *> ?: return emptyMap()
            return buildMap {
                for ((key, value) in map) {
                    val convoyId = key as? String ?: continue
                    val millis = (value as? Timestamp)?.toDate()?.time ?: continue
                    put(convoyId, millis)
                }
            }
        }
        private const val CONVOY_LIST = "convoy-list"
        private const val POST = "convoyChat-post"
        private const val LIST = "convoyChat-list"
        private const val MARK_READ = "convoyChat-markRead"

        /**
         * Newest-message window scanned for the unread count (see observeUnread).
         * One more than ConvoyBar.UNREAD_DISPLAY_MAX, so the count saturates
         * exactly where the badge starts showing "9+".
         */
        private const val UNREAD_SCAN_LIMIT = 10L

        fun createIfAvailable(context: Context): ConvoyChatRepository? {
            if (FirebaseApp.getApps(context).isEmpty()) return null
            return FirebaseConvoyChatRepository(
                FirebaseFirestore.getInstance(),
                FirebaseFunctions.getInstance(CHANNEL_FUNCTIONS_REGION),
                FirebaseBlockVisibilityRepository.createOrEmpty(context),
                FirebaseLiveProfileRepository.sharedOrEmpty(context),
            )
        }
    }
}

File: apps/ios/KCC/Chat/FirebaseConvoyChatRepository.swift
import FirebaseAuth
import FirebaseCore
import FirebaseFirestore
import FirebaseFunctions
import Foundation

/// ``ConvoyChatRepository`` backed by the member-readable
/// `convoyChats/{convoyId}/messages` listener plus the member-gated `convoy-list`
/// / `convoyChat-*` / `chatchannels-*` callables (europe-west1) — the iOS port of
/// Android's `FirebaseConvoyChatRepository`.
///
/// The convoy list comes from `convoy-list`, projected client-side to
/// accepted-member convoys (``ConvoyChatMapper``). Each convoy's newest window is
/// a bounded live listener block-filtered against the caller's mutual-hidden set;
/// older pages come from `convoyChat-list`. The client never writes the message
/// tree — `convoyChat-post` stamps authorship and the denormalized profile.
///
/// Construction is guarded (``createIfAvailable()`` returns nil without config).
final class FirebaseConvoyChatRepository: ConvoyChatRepository, @unchecked Sendable {
    private let firestore: Firestore
    private let functions: Functions

    private init(firestore: Firestore, functions: Functions) {
        self.firestore = firestore
        self.functions = functions
    }

    func listConvoys() async -> ConvoyListState {
        do {
            let data = try await functions.httpsCallable(Self.convoyListCallable).call([String: Any]()).data
            return .loaded(ConvoyChatMapper.chatEligibleConvoys(data))
        } catch {
            // Transient callable failure — retryable.
            return .error
        }
    }

    func observeMessages(convoyId: String) -> AsyncStream<ChannelMessagesState> {
        let query = firestore
            .collection(Self.convoyChatsCollection)
            .document(convoyId)
            .collection(Self.messagesCollection)
            .order(by: Self.createdAtField, descending: true)
            .limit(to: channelMessagesPageSize)
        let blockDoc = currentUserId().map {
            firestore.collection(ChatFirestore.blockVisibilityCollection).document($0)
        }
        return ChannelMessagesListener.stream(messagesQuery: query, blockVisibilityDocument: blockDoc)
    }

    func loadOlder(convoyId: String, before: String) async -> ChannelOlderResult {
        do {
            let data = try await functions.httpsCallable(Self.listCallable)
                .call(["convoyId": convoyId, "before": before]).data
            return .loaded(ChannelResponseParser.parseMessagesPage(data))
        } catch {
            return .failed
        }
    }

    func post(
        convoyId: String,
        text: String,
        clientId: String?,
        replyToMessageId: String?
    ) async -> ChannelSendResult {
        // Convoy chat accepts no @mentions (the field exists only so both
        // channels share one message shape).
        var payload: [String: Any] = [
            "convoyId": convoyId,
            "text": text.trimmingCharacters(in: .whitespacesAndNewlines),
        ]
        if let clientId { payload["clientId"] = clientId }
        if let replyToMessageId { payload["replyToMessageId"] = replyToMessageId }
        do {
            let data = try await functions.httpsCallable(Self.postCallable).call(payload).data
            return ChannelResponseParser.parsePostSuccess(data)
        } catch {
            return .failed(ChannelErrorMapper.mapSend(ChatFirestore.channelErrorCode(from: error)))
        }
    }

    func markRead(convoyId: String) async {
        _ = try? await functions.httpsCallable(Self.markReadCallable).call(["convoyId": convoyId])
    }

    func report(convoyId: String, messageId: String, reason: ChatReportReason) async -> ChannelReportResult {
        // `channel: "convoy"` REQUIRES convoyId (the backend rejects it missing);
        // the pair mirrors the convoy read gate.
        let payload: [String: Any] = [
            "channel": "convoy",
            "convoyId": convoyId,
            "messageId": messageId,
            "reason": reason.wire,
        ]
        do {
            _ = try await functions.httpsCallable(Self.reportCallable).call(payload)
            return .reported
        } catch {
            return .failed
        }
    }

    func currentUserId() -> String? {
        Auth.auth().currentUser?.uid
    }

    // MARK: - Constants

    private static let convoyChatsCollection = "convoyChats"
    private static let messagesCollection = "messages"
    private static let createdAtField = "createdAt"
    private static let convoyListCallable = "convoy-list"
    private static let postCallable = "convoyChat-post"
    private static let listCallable = "convoyChat-list"
    private static let markReadCallable = "convoyChat-markRead"
    private static let reportCallable = "chatchannels-reportMessage"

    // MARK: - Factory

    private static let cachedLock = NSLock()
    nonisolated(unsafe) private static var cached: FirebaseConvoyChatRepository?

    /// Returns the process-wide repository when Firebase is configured, or nil in
    /// a config-less build. Emulator seams as in ``FirebaseCommunityChatRepository``.
    static func createIfAvailable() -> ConvoyChatRepository? {
        guard FirebaseApp.app() != nil else { return nil }
        cachedLock.lock()
        defer { cachedLock.unlock() }
        if let cached { return cached }
        let firestore = Firestore.firestore()
        if let emulator = FirebaseEmulatorHost.parse(
            ProcessInfo.processInfo.environment["FIREBASE_FIRESTORE_EMULATOR_HOST"]
        ) {
            firestore.useEmulator(withHost: emulator.host, port: emulator.port)
        }
        let functions = Functions.functions(region: ChatFirestore.functionsRegion)
        if let emulator = FirebaseEmulatorHost.parse(
            ProcessInfo.processInfo.environment["FIREBASE_FUNCTIONS_EMULATOR_HOST"]
        ) {
            functions.useEmulator(withHost: emulator.host, port: emulator.port)
        }
        let repository = FirebaseConvoyChatRepository(firestore: firestore, functions: functions)
        cached = repository
        return repository
    }
}
